### PR TITLE
#1801 Using constructor as builderCreationMethod in custom builder

### PR DESCRIPTION
**Pull request by sjaakd**

### CODE REVIEW

**Comment by sjaakd** [comment]
test this please

**Comment by sjaakd** [comment]
See also #1809

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import javax.lang.model.util.ElementFilter;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import org.mapstruct.ap.spi.BuilderInfo;
+import org.mapstruct.ap.spi.BuilderProvider;
+import org.mapstruct.ap.spi.ImmutablesBuilderProvider;
+import org.mapstruct.ap.spi.MoreThanOneBuilderCreationMethodException;
+
```
Comment (Captain1653) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java]: Maybe javadoc @author?

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            return null;
+        }
+
+        List<TypeElement> types = ElementFilter.typesIn( typeElement.getEnclosedElements() );
+        List<BuilderInfo> builderInfo = new ArrayList<>();
+        types.stream()
+             .filter( innerType -> isPossibleInnerBuilder( innerType, typeElement ) )
+             .forEach( innerBuilderClass -> {
+                 Collection<ExecutableElement> buildMethods = findBuildMethods( innerBuilderClass, typeElement );
+                 if (!buildMethods.isEmpty()) {
```
Comment (Captain1653) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java]: Maybe is it better to refactor it? :)  It's little hard to understand

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            throw new MoreThanOneBuilderCreationMethodException(typeElement.asType(), builderInfo);
+        }
+
+        return null;
+    }
+
+    private boolean isPossibleInnerBuilder(TypeElement innerClass, TypeElement outerClass) {
+        final String packageName = this.elementUtils.getPackageOf( outerClass ).getQualifiedName().toString();
+        final Name outerSimpleName = outerClass.getSimpleName();
+        final String builderClassName = packageName + ".Immutable" + outerSimpleName + ".Builder";
```
Comment (Captain1653) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java]: Do we really need modificator  "final" for variables: packageName, outerSimpleName, builderClassName?

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java (hunk 1 above)
Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java]: done

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        return innerClass.getSimpleName().contentEquals( "Builder" ) &&
+            getTypeElement( innerClass.getSuperclass() ).getQualifiedName().contentEquals( builderClassName ) &&
+            innerClass.getModifiers().contains( Modifier.PUBLIC );
+    }
+
+    @Override
+    protected BuilderInfo findBuilderInfo(TypeElement typeElement) {
+        Name name = typeElement.getQualifiedName();
+        if ( name.toString().endsWith( ".Item" ) ) {
+            BuilderInfo info = findBuilderInfoForImmutables( typeElement );
```
Comment (Captain1653) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java]: Maybe it's better to do something like that ?  :
`  Name name = typeElement.getQualifiedName();
  BuilderInfo info;
 if ( name.toString().endsWith( ".Item" ) ) {
        info = findBuilderInfoForImmutables( typeElement );
 }
 return info == null ? super.findBuilderInfo(typeElement) : info;


`

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/ItemMapper.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1801;
+
+import org.mapstruct.Builder;
+import org.mapstruct.Mapper;
+import org.mapstruct.ap.test.bugs._1801.domain.Item;
+import org.mapstruct.ap.test.bugs._1801.dto.ItemDTO;
+import org.mapstruct.factory.Mappers;
+
```
Comment (Captain1653) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/ItemMapper.java]: Maybe javadoc @author?

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/domain/ImmutableItem.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1801.domain;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
```
Comment (Captain1653) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/domain/ImmutableItem.java]: Maybe javadoc @author?

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/domain/ImmutableItem.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+ * Immutable implementation of {@link Item}.
+ * <p>
+ * Superclass should expose a static subclass of the Builder to create immutable instance
+ * {@code public static Builder extends ImmutableItem.Builder}.
+ */
+@SuppressWarnings({"all"})
```
Comment (Captain1653) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/domain/ImmutableItem.java]: @SuppressWarnings({"all}) ? It is really necessary? Maybe should we exclude some part?

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                                      constructor.getModifiers().contains( Modifier.PUBLIC ) )
+                                  .findAny()
+                                  .ifPresent( defaultConstructor -> builderInfo.add(
+                                      new BuilderInfo.Builder()
+                                          .builderCreationMethod( defaultConstructor )
+                                          .buildMethod( buildMethods )
+                                          .build()
+                                  ) );
+                 }
+             } );
```
[the remark this answers is not in the document]
Comment (filiphr) [reply processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java]: This is a test code so I would say that it is OK to leave it as is.

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/domain/ImmutableItem.java (hunk 7 above)
Comment (filiphr) [reply processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/domain/ImmutableItem.java]: This class is identical to the one generated by `Immutables`. That's why the supress is there I think.

**Comment by filiphr** [comment]
LGTM @sjaakd. Will it make sense to squash some commits? Perhaps the first 3 into one authored by @dengzhizhi and the last 2 into one authored by you @sjaakd?

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/domain/ImmutableItem.java (hunk 7 above)
Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/domain/ImmutableItem.java]: did some polishing.. It still captures the essence but fixes suppress warnings

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java]: went ahead and polished it anyway. hope its better now

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java (hunk 4 above)
Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/bugs/_1801/Issue1801BuilderProvider.java]: I left this as is..

**Comment by sjaakd** [comment]
@Captain1653 , @filiphr thanks for reviewing..

**Pull request merged**